### PR TITLE
[FLINK-19070][hive] Hive connector should throw a meaningful exception if user reads/writes ACID tables

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Hive ACID table configures tblproperties with 'transactional=true'. Currently Hive connector couldn't support to read and write on ACID tables. Therefore, Hive connector should throw a meaningful exception to users for reading or writing ACID tables.*

## Brief change log

  - *`HiveTableSource#getDataStream` adds `checkAcidTable` to check whether the tblproperties `transactional` of the Hive source table is true. If the tblproperties `transactional` is true, this should throw `FlinkHiveException` to prompt user meaningful exception.*
 - *`HiveTableSink#consumeDataStream` adds `checkAcidTable` to check whether the tblproperties `transactional` of the Hive sink table is true. If the tblproperties `transactional` is true, this should throw `FlinkHiveException` to prompt user meaningful exception.*
 - *Add `HiveUtils#checkAcidTable` which utils for Hive connector to check whether to read or write on the hive ACID table.*

## Verifying this change

  - *Add `TableEnvHiveConnectorITCase#testStreamTransactionalTable` and `TableEnvHiveConnectorITCase#testBatchTransactionalTable`to verify the test case that create hive table with the tblproperties `transactional=true` for the meaningful exception message.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by lirui-apache** [comment]
@SteNicholas Thanks for working on this. I think the check should be enforced when getting source/sink for a hive table, rather than creating the hive table itself.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java [hunk 1]
```diff
@@ -328,4 +330,16 @@ public void setStaticPartition(Map<String, String> partitionSpec) {
...
 	}
+
+	private void checkAcidTable() {
+		if (catalogTable != null && catalogTable.getOptions() != null) {
+			String tableIsTransactional = catalogTable.getOptions().get("transactional");
+			if (tableIsTransactional == null) {
+				tableIsTransactional = catalogTable.getOptions().get("transactional".toUpperCase());
+			}
+			if (tableIsTransactional != null && tableIsTransactional.equalsIgnoreCase("true")) {
+				throw new TableException(String.format("Cannot write on the ACID table %s.", identifier.asSummaryString()));
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java:341]:
```suggestion
				throw new FlinkHiveException(String.format("Writing ACID table %s is not supported", identifier.asSummaryString()));
```

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 2]
```diff
@@ -556,4 +558,16 @@ public String explainSource() {
...
 	}
+
+	private void checkAcidTable() {
+		if (catalogTable != null && catalogTable.getOptions() != null) {
+			String tableIsTransactional = catalogTable.getOptions().get("transactional");
+			if (tableIsTransactional == null) {
+				tableIsTransactional = catalogTable.getOptions().get("transactional".toUpperCase());
+			}
+			if (tableIsTransactional != null && tableIsTransactional.equalsIgnoreCase("true")) {
+				throw new TableException(String.format("Cannot read on the ACID table %s.", tablePath));
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:569]:
```suggestion
				throw new FlinkHiveException(String.format("Reading ACID table %s is not supported", tablePath));
```

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java [hunk 3]
```diff
@@ -330,6 +332,18 @@ public void setOverwrite(boolean overwrite) {
 		this.overwrite = overwrite;
 	}
 
+	private void checkAcidTable() {
+		if (catalogTable != null && catalogTable.getOptions() != null) {
+			String tableIsTransactional = catalogTable.getOptions().get("transactional");
+			if (tableIsTransactional == null) {
+				tableIsTransactional = catalogTable.getOptions().get("transactional".toUpperCase());
+			}
+			if (tableIsTransactional != null && tableIsTransactional.equalsIgnoreCase("true")) {
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java:340]: Let's extract these into a util method

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorITCase.java [hunk 4]
```diff
@@ -574,6 +575,43 @@ public void testStreamCompressTextTable() throws Exception {
...
+			getTableEnvWithHiveCatalog() :
+			getStreamTableEnvWithHiveCatalog();
+		tableEnv.executeSql("create database db1");
+		try {
+			tableEnv.executeSql("create table db1.src (x string,y string)");
+			hiveShell.execute("create table db1.dest (x string,y string) clustered by (x) into 3 buckets stored as orc tblproperties ('transactional'='true')");
+			HiveTestUtils.createTextTableInserter(hiveShell, "db1", "src")
+				.addRow(new Object[]{"a", "b"})
+				.addRow(new Object[]{"c", "d"})
+				.commit();
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/TableEnvHiveConnectorITCase.java:589]: No need to insert data

**Review by lirui-apache — APPROVED** [review]
+1, LGTM

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java [hunk 5]
```diff
@@ -403,6 +404,25 @@ public static Table instantiateHiveTable(ObjectPath tablePath, CatalogBaseTable
...
 
+	/**
+	 * Check whether to read or write on the hive ACID table.
+	 *
+	 * @param catalogTable Hive catalog table.
+	 * @param tablePath    Identifier table path.
+	 * @throws FlinkHiveException Thrown, if the source or sink table is transactional.
+	 */
+	public static void checkAcidTable(CatalogTable catalogTable, ObjectPath tablePath) {
+		if (catalogTable != null && catalogTable.getOptions() != null) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java:415]: `catalogTable` and `tablePath` never null, you can take a look to the constructor of `HiveTableSource`

**Review by JingsongLi — APPROVED** [review]
Looks good to me.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java (hunk 5 above)
Comment (SteNicholas) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java:415]: @JingsongLi `checkAcidTable` is the util method for `HiveTableUtil`, which should add check for the parameter `catalogTable`.

Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java:415]: If really want to have a check, check should be `Preconditions.check`

Comment (SteNicholas) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java]: @JingsongLi Yeah, you are right. I remove this check for `catalogTable` with options.

**Comment by JingsongLi** [comment]
Merging...

**Pull request merged**